### PR TITLE
Tdl 18019 upgrade sdk api version

### PR DESCRIPTION
**Pull request by prijendev**
# Description of change
**[TDL 18019] Upgrade SDK and API version**
- Removed deprecated stream `bidding_rules`  and it's a schema
- Added new stream `tracking_tags`.
- Added new field `bid_strategy` in the schema of `line_items` stream.

**[TLD 18090] Resolves json-schema $refs with the appropriate dict.**
- Replaced custom resolve_schema_references with standard singer method.

**Note**: Removed non-standard test cases as we have separate https://github.com/singer-io/tap-twitter-ads/pull/20 for standard tap tester.
# Manual QA steps
 - 
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

**Review by savan-chovatiya — APPROVED** [review]

**Review by KrisPersonal — APPROVED** [review]

**Review by dbshah1212 — APPROVED** [review]

File: tests/test_configuration.py [hunk 1]
```diff
@@ -1,87 +0,0 @@
-def config():
-    return {
```
Comment (prijendev) [inline tests/test_configuration.py:2]: Removed test cases as we have separate [PR#20](https://github.com/singer-io/tap-twitter-ads/pull/20) for standard tap tester.

File: tap_twitter_ads/schema.py [hunk 2]
```diff
@@ -103,10 +107,13 @@ def get_schemas(reports):
...
         mdata = metadata.get_standard_metadata(
             schema=schema,
-            key_properties=stream_metadata.get('key_properties', None),
-            valid_replication_keys=stream_metadata.get('replication_keys', None),
-            replication_method=stream_metadata.get('replication_method', None)
+            key_properties= (hasattr(stream_metadata, 'key_properties') or None) and stream_metadata.key_properties,
+            valid_replication_keys=(hasattr(stream_metadata, 'replication_keys') or None) and stream_metadata.replication_keys,
+            replication_method=(hasattr(stream_metadata, 'replication_method') or None) and stream_metadata.replication_method
         )
+        # make replication keys of automatic inclusion
```
Comment (savan-chovatiya) [inline tap_twitter_ads/schema.py:114]:
```suggestion
        # Make replication keys of automatic inclusion
```

File: tap_twitter_ads/schema.py [hunk 3]
```diff
@@ -193,6 +200,10 @@ def get_schemas(reports):
             valid_replication_keys=['end_time'],
             replication_method='INCREMENTAL'
         )
+
+        # make replication keys of automatic inclusion
```
Comment (savan-chovatiya) [inline tap_twitter_ads/schema.py:204]:
```suggestion
        # Make replication keys of automatic inclusion
```

File: tap_twitter_ads/sync.py [hunk 4]
```diff
@@ -1021,11 +36,14 @@ def sync(client, config, catalog, state):
...
-    flat_streams = flatten_streams()
+    # flat_streams = flatten_streams()
     # Loop thru all streams
-    for stream_name, stream_metadata in flat_streams.items():
+    
+    # for stream_name, stream_metadata in flat_streams.items():
+    for stream_name, stream_obj in STREAMS.items():
         # If stream has a parent_stream, then it is a child stream
-        parent_stream = stream_metadata.get('parent_stream')
+        # parent_stream = stream_obj.get('parent_stream')
```
Comment (savan-chovatiya) [inline tap_twitter_ads/sync.py:45]: Remove these commented lines

File: tap_twitter_ads/sync.py [hunk 5]
```diff
@@ -1119,15 +140,15 @@ def sync(client, config, catalog, state):
 
                 LOGGER.info('Report: {} - START Syncing for Account ID: {}'.format(
                     report_name, account_id))
-
+
```
Comment (savan-chovatiya) [inline tap_twitter_ads/sync.py:143]:
```suggestion
 
```

**Review by savan-chovatiya — CHANGES_REQUESTED** [review]

File: tap_twitter_ads/sync.py (hunk 4 above)
Comment (prijendev) [reply tap_twitter_ads/sync.py]: Updated

File: tap_twitter_ads/schema.py (hunk 3 above)
Comment (prijendev) [reply tap_twitter_ads/schema.py]: Updated

File: tap_twitter_ads/schema.py (hunk 2 above)
Comment (prijendev) [reply tap_twitter_ads/schema.py]: Updated

File: tap_twitter_ads/schema.py [hunk 6]
```diff
@@ -103,10 +107,13 @@ def get_schemas(reports):
         # https://github.com/singer-io/singer-python/blob/master/singer/metadata.py#L25-L44
         mdata = metadata.get_standard_metadata(
             schema=schema,
-            key_properties=stream_metadata.get('key_properties', None),
-            valid_replication_keys=stream_metadata.get('replication_keys', None),
-            replication_method=stream_metadata.get('replication_method', None)
+            key_properties= (hasattr(stream_metadata, 'key_properties') or None) and stream_metadata.key_properties,
```
Comment (dbshah1212) [inline tap_twitter_ads/schema.py:110]: instead of this use `getattr(a, 'property', 'default value')`

Comment (dbshah1212) [reply tap_twitter_ads/schema.py:110]: Also updated this for each hasattr

**Review by kspeer825 — APPROVED** [review]
The tap-tester tests are out of date with a few of the other feature branches. Approving, assuming these will be overridden by those other branches.

File: tests/unittests/test_page_size.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+        self.assertEqual(actual_value, expected_value)
+
+    def test_negative_float_page_size_field(self):
+        """ Verify that page_size is set to 100 if negative float is given in the config """
+
+        expected_value = DEFAULT_PAGE_SIZE
+        actual_value = get_page_size(get_config(-10.5), DEFAULT_PAGE_SIZE)
+
+        self.assertEqual(actual_value, expected_value)
```
Comment (RushT007) [inline tests/unittests/test_page_size.py:85]: These should be invalid and we should raise exceptions.

**Review by RushT007 — CHANGES_REQUESTED** [review]

File: tests/unittests/test_page_size.py (hunk 7 above)
Comment (prijendev) [reply tests/unittests/test_page_size.py]: Updated `page_size` validation logic.

File: tests/test_twitter_ads_start_date.py [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+
+        # run initial sync
+        record_count_by_stream_1 = self.run_and_verify_sync(conn_id_1)
+        synced_records_1 = runner.get_records_from_target_output()
+
+        ##########################################################################
+        # Update START DATE Between Syncs
+        ##########################################################################
+   
+        print("REPLICATION START DATE CHANGE: {} ===>>> {} ".format(
```
Comment (RushiT0122) [inline tests/test_twitter_ads_start_date.py:81]: Can we remove `print` statement and use `tap_tester.logger.LOGGER` instead?

**Review by RushiT0122 — CHANGES_REQUESTED** [review]
Minor change in start date integration test, otherwise everything looks good.

File: tests/test_twitter_ads_start_date.py (hunk 8 above)
Comment (prijendev) [reply tests/test_twitter_ads_start_date.py]: Replaced print statement with `LOGGER`.

**Review by RushiT0122 — CHANGES_REQUESTED** [review]

File: tests/test_twitter_ads_start_date.py [hunk 9]
```diff
@@ -0,0 +1,180 @@
+import singer
+import tap_tester.connections as connections
+import tap_tester.runner as runner
+import tap_tester.menagerie as menagerie
+
+LOGGER = singer.get_logger()
```
[the remark this answers is not in the document]
Comment (prijendev) [reply tests/test_twitter_ads_start_date.py]: Replaced singer logger with tap_tester logger.

**Review by KrisPersonal — APPROVED** [review]

**Review by RushiT0122 — APPROVED** [review]

**Pull request merged**